[webgpu] Add Surface API

beaufortfrancois opened the pull request:
FIX https://github.com/emscripten-core/emscripten/issues/21745

beaufortfrancois commented on src/library_webgpu.js:
```diff
@@ -2673,13 +2688,89 @@ var LibraryWebGPU = {
+  wgpuSurfaceGetCapabilities: (surfaceId, adapterId, capabilitiesPtr) => {
+    if (capabilitiesPtr !== 0) {
+      {{{ makeSetValue('capabilitiesPtr', C_STRUCTS.WGPUSurfaceCapabilities.formatCount, '1', 'i32') }}};
+      var formatsPtr = WebGPU.TextureFormat.indexOf("bgra8unorm"); // FIXME
```
@kainino0x Do you know how I can properly fill `formats` in `wgpuSurfaceGetCapabilities()`?

beaufortfrancois commented on src/library_webgpu.js:
```diff
@@ -2673,13 +2688,90 @@ var LibraryWebGPU = {
+    if (capabilitiesPtr !== 0) {
+      {{{ makeSetValue('capabilitiesPtr', C_STRUCTS.WGPUSurfaceCapabilities.formatCount, '1', 'i32') }}};
+      var format = navigator["gpu"]["getPreferredCanvasFormat"]();
+      var formatsPtr = WebGPU.Int_PreferredFormat[format]; // FIXME
```
@kainino0x Do you know how I can properly fill `formats` in `wgpuSurfaceGetCapabilities()`?

kainino0x reviewed (commented):
Sorry haven't finished reviewing but here are some comments

kainino0x commented on src/library_webgpu.js:
```diff
@@ -2673,13 +2688,90 @@ var LibraryWebGPU = {
+  wgpuSurfaceGetCurrentTexture: (surfaceId, surfaceTexturePtr) => {
+    var context = WebGPU.mgrSurface.get(surfaceId);
+    var texture = WebGPU.mgrTexture.create(context.getCurrentTexture());
+    if (surfaceTexturePtr !== 0) {
```
This pointer is non-nullable so you can `makeCheck` that it's non-null.

beaufortfrancois replied:
Got it!

kainino0x commented on src/library_webgpu.js:
```diff
@@ -2673,13 +2688,90 @@ var LibraryWebGPU = {
+    if (surfaceTexturePtr !== 0) {
+      {{{ makeSetValue('surfaceTexturePtr', C_STRUCTS.WGPUSurfaceTexture.texture, 'texture', '*') }}};
+      {{{ makeSetValue('surfaceTexturePtr', C_STRUCTS.WGPUSurfaceTexture.suboptimal, '0', 'i32') }}};
+      {{{ makeSetValue('surfaceTexturePtr', C_STRUCTS.WGPUSurfaceTexture.status, '0', 'i32') }}};
```
Need to catch if getCurrentTexture throws an exception and return that in the status.

beaufortfrancois replied:
Note that getCurrentTexture() can throw a JavaScript error if the context is not configured but there's no matching error for this. So I'm using `1` for now but that's not good ;(

beaufortfrancois commented on src/library_webgpu.js:
```diff
@@ -99,6 +99,13 @@ wgpu${type}Release: (id) => WebGPU.mgr${type}.release(id),`;
+      'premultiplied',
+      'unpremultiplied',
+      'inherit',
+    ],
```
Done

kainino0x replied:
My bad, this is needed. I saw that it wasn't used and didn't think about the fact that it _should_ be used.

However not all of the entries should be there, it should be something like:
```
    CompositeAlphaMode: [
      'opaque',
      'opaque',
      'premultiplied',
    ],
```

beaufortfrancois commented on src/library_webgpu.js:
```diff
@@ -568,6 +575,14 @@ var LibraryWebGPU = {
+      'lost',
+      'out-of-memory',
+      'device-lost',
+    ],
```
Done

beaufortfrancois commented on src/library_webgpu.js:
```diff
@@ -2673,13 +2688,90 @@ var LibraryWebGPU = {
+      // "presentModes",
+      // "alphaModeCount",
+      // "alphaModes"
+    }
```
Thank you! That was the missing part for me.

beaufortfrancois commented:
> Sorry haven't finished reviewing but here are some comments

Thanks for starting the review @kainino0x.
I've addressed all your feedback.

There are still some TODOs I'd like to solve such as presentModes, alphaModes, and the error returned by getCurrentTexture.

kainino0x commented on system/lib/webgpu/webgpu.cpp:
```diff
@@ -27,3 +38,24 @@ WGPUInstance wgpuCreateInstance(const WGPUInstanceDescriptor* descriptor) {
+        kRGBA8UnormPreferredContextFormats.data());
+  }
+
+  // TODO: What do we return for presentModes and alphaModes?
```
presentModes should statically be `[Fifo]`.

alphaModes should statically be one of the following, but I'm not sure which (https://github.com/webgpu-native/webgpu-headers/issues/292):
- `[Auto, Opaque, Premultiplied]`
- `[Opaque, Premultiplied]` I would pick this one for now.

kainino0x commented on test/webgpu_basic_rendering.cpp:
```diff
@@ -383,15 +391,21 @@ void run() {
+            .alphaMode = capabilities.alphaModes[0],
+            .width = kWidth,
+            .height = kHeight,
+            // .presentMode = capabilities.presentModes[0]};
+            .presentMode = wgpu::PresentMode::Fifo};
```
Should always use `Fifo` (it's guaranteed to be available).
```suggestion
            .presentMode = wgpu::PresentMode::Fifo};
```

kainino0x commented on test/webgpu_basic_rendering.cpp:
```diff
@@ -383,15 +391,21 @@ void run() {
+            .device = device,
+            .format = capabilities.formats[0],
+            .usage = wgpu::TextureUsage::RenderAttachment,
+            .alphaMode = capabilities.alphaModes[0],
```
Should always use `Auto` (it's guaranteed to be available).
```suggestion
            .alphaMode = wgpu::CompositeAlphaMode::Auto,
```

kainino0x commented on system/lib/webgpu/webgpu.cpp:
```diff
@@ -27,3 +38,24 @@ WGPUInstance wgpuCreateInstance(const WGPUInstanceDescriptor* descriptor) {
+         preferredFormat == WGPUTextureFormat_RGBA8Unorm);
+  capabilities->formatCount = 3;
+  if (preferredFormat == WGPUTextureFormat_RGBA8Unorm) {
+    capabilities->formats = reinterpret_cast<const WGPUTextureFormat*>(
```
We should try to avoid reinterpret_cast. Why is it needed? I think `.data()` should already have the right type.

kainino0x commented on system/lib/webgpu/webgpu.cpp:
```diff
@@ -9,8 +9,19 @@
+static constexpr std::array<WGPUTextureFormat, 3>
+    kRGBA8UnormPreferredContextFormats = {WGPUTextureFormat_RGBA8Unorm,
+                                          WGPUTextureFormat_BGRA8Unorm,
+                                          WGPUTextureFormat_RGBA16Float};
```
nit: Define these inside `wgpuSurfaceGetCapabilities` since they're only used there. (They're static so you can still return pointers to them.)

kainino0x commented on system/lib/webgpu/webgpu.cpp:
```diff
@@ -27,3 +38,24 @@ WGPUInstance wgpuCreateInstance(const WGPUInstanceDescriptor* descriptor) {
+
+void wgpuSurfaceGetCapabilities(WGPUSurface surface,
+                                WGPUAdapter adapter,
+                                WGPUSurfaceCapabilities* capabilities) {
```
Assert `capabilities->nextInChain == nullptr` because we don't support any extensions and GetCapabilities doesn't have a return code yet (it will, later).
```suggestion
                                WGPUSurfaceCapabilities* capabilities) {
  assert(capabilities->nextInChain == nullptr); // TODO: Return WGPUStatus_Error
```

kainino0x commented on system/lib/webgpu/webgpu.cpp:
```diff
@@ -27,3 +38,24 @@ WGPUInstance wgpuCreateInstance(const WGPUInstanceDescriptor* descriptor) {
+      wgpuSurfaceGetPreferredFormat(surface, adapter);
+  assert(preferredFormat == WGPUTextureFormat_BGRA8Unorm ||
+         preferredFormat == WGPUTextureFormat_RGBA8Unorm);
+  capabilities->formatCount = 3;
```
nit: move this inside the `if` branches and use `.size()` so that the formatCount will be correct by construction.

kainino0x commented on src/library_webgpu.js:
```diff
@@ -2673,13 +2673,84 @@ var LibraryWebGPU = {
+#if ASSERTIONS
+      err(`wgpuSurfaceGetCurrentTexture() failed: ${ex}`);
+#endif
+      {{{ makeSetValue('surfaceTexturePtr', C_STRUCTS.WGPUSurfaceTexture.status, /*Timeout=*/1, 'i32') }}};
```
I filed https://github.com/webgpu-native/webgpu-headers/issues/291. For now, let's use `DeviceLost` since it's sorta appropriate (if it's unconfigured, there's no device at all).

```suggestion
      // TODO(https://github.com/webgpu-native/webgpu-headers/issues/291): What should the status be here?
      {{{ makeSetValue('surfaceTexturePtr', C_STRUCTS.WGPUSurfaceTexture.status,
        gpu.SurfaceGetCurrentTextureStatus.DeviceLost, 'i32') }}};
```

kainino0x commented on src/library_webgpu.js:
```diff
@@ -2673,13 +2673,84 @@ var LibraryWebGPU = {
+      var texture = WebGPU.mgrTexture.create(context.getCurrentTexture());
+      {{{ makeSetValue('surfaceTexturePtr', C_STRUCTS.WGPUSurfaceTexture.texture, 'texture', '*') }}};
+      {{{ makeSetValue('surfaceTexturePtr', C_STRUCTS.WGPUSurfaceTexture.suboptimal, '0', 'i32') }}};
+      {{{ makeSetValue('surfaceTexturePtr', C_STRUCTS.WGPUSurfaceTexture.status, /*Success=*/0, 'i32') }}};
```
Define a new table in the `Compile-time table for enum integer values used with templating` section:

```js
    SurfaceGetCurrentTextureStatus: {
      Success: 0,
      DeviceLost: 5,
    },
```

so you can name these explicitly:

```suggestion
      {{{ makeSetValue('surfaceTexturePtr', C_STRUCTS.WGPUSurfaceTexture.status, 
        gpu.SurfaceGetCurrentTextureStatus.Success, 'i32') }}};
```

kainino0x commented on system/lib/webgpu/webgpu.cpp:
```diff
@@ -27,3 +38,25 @@ WGPUInstance wgpuCreateInstance(const WGPUInstanceDescriptor* descriptor) {
+  assert(preferredFormat == WGPUTextureFormat_BGRA8Unorm ||
+         preferredFormat == WGPUTextureFormat_RGBA8Unorm);
+  capabilities->formatCount = 3;
+  if (preferredFormat == WGPUTextureFormat_RGBA8Unorm) {
```
Also just noticed this condition is reversed.

kainino0x reviewed (commented):
I addressed all my own comments, so LGTM now, but @beaufortfrancois PTAL before I land it.

beaufortfrancois commented:
> I addressed all my own comments, so LGTM now, but @beaufortfrancois PTAL before I land it.

@kainino0x I've just merged main branch and I've tried locally all your latest changes. LGTM. Thank you!

juj commented:
Is there an IDL or a spec available to this Surface API? It doesn't seem to exist in the [WebGPU spec](https://www.w3.org/TR/webgpu/#idl-index) at least.

beaufortfrancois commented:
> Is there an IDL or a spec available to this Surface API? It doesn't seem to exist in the [WebGPU spec](https://www.w3.org/TR/webgpu/#idl-index) at least.

I believe https://github.com/webgpu-native/webgpu-headers/pull/164 is where things got defined.

kainino0x commented:
> Is there an IDL or a spec available to this Surface API? It doesn't seem to exist in the [WebGPU spec](https://www.w3.org/TR/webgpu/#idl-index) at least.

The Web API only concerns itself with HTML Canvas, whereas Surface is a C-API-only abstraction over both HTML Canvas and native window system integration. The Surface API is very close to [GPUCanvasContext](https://gpuweb.github.io/gpuweb/#canvas-context) though.

juj commented:
What got me confused is that the JS side code `configure`s a GPUCanvasContext using a `presentMode` parameter here: https://github.com/emscripten-core/emscripten/pull/21939/files#diff-f2ba55d634491c159f57989d779458c308f77253f42039012674780ff95894daR2716

but there is no such parameter defined in the IDL: https://www.w3.org/TR/webgpu/#dictdef-gpucanvasconfiguration (and no `presentMode` mentioned in the spec page). Is that something that is not yet defined in the WebGPU spec, or maybe a typo leaking over from the Dawn side?

kainino0x commented:
Ah, that's not supposed to be there. presentMode is native-only, Chrome doesn't support it in Blink, it just gets dropped on the floor because that's how WebIDL works.

kainino0x commented on src/library_webgpu.js:
```diff
@@ -2676,11 +2680,81 @@ var LibraryWebGPU = {
+      // viewFormatCount
+      // viewFormats
+      "alphaMode": "opaque",
+      "width": canvasSize[0],
+      "height": canvasSize[1],
+      "presentMode": {{{ gpu.makeGetU32('config', C_STRUCTS.WGPUSurfaceConfiguration.presentMode) }}},
```
These don't exist in the JS API.
```suggestion
```

beaufortfrancois replied:
Good call!

kainino0x commented on src/library_webgpu.js:
```diff
@@ -2676,11 +2680,81 @@ var LibraryWebGPU = {
+      "usage": {{{ gpu.makeGetU32('config', C_STRUCTS.WGPUSurfaceConfiguration.usage) }}},
+      // viewFormatCount
+      // viewFormats
+      "alphaMode": "opaque",
```
This should be either `"opaque"` or `"premultiplied"` depending on the configuration, and there should be an assert that it's one of the values that's valid in JS (auto, opaque, premultiplied). (Just look it up in the table above and then assert that the result is not `undefined`.)

Alternatively, assert the alphaMode is Opaque for now with a TODO, and implement this later if you'd like.

beaufortfrancois replied:
I've assert alphaMode is Opaque with a TODO.

kainino0x commented on src/library_webgpu.js:
```diff
@@ -2676,11 +2680,81 @@ var LibraryWebGPU = {
+        {{{ gpu.makeGetU32('config', C_STRUCTS.WGPUSurfaceConfiguration.format) }}}],
+      "usage": {{{ gpu.makeGetU32('config', C_STRUCTS.WGPUSurfaceConfiguration.usage) }}},
+      // viewFormatCount
+      // viewFormats
```
Assert `viewFormatCount === 0 && viewFormats === 0` since we don't support them yet (with "TODO" in the assert message).

beaufortfrancois replied:
Done

kainino0x commented on test/webgpu_basic_rendering.cpp:
```diff
@@ -400,7 +400,7 @@ void run() {
             .format = capabilities.formats[0],
             .usage = wgpu::TextureUsage::RenderAttachment,
-            .alphaMode = wgpu::CompositeAlphaMode::Auto,
+            .alphaMode = wgpu::CompositeAlphaMode::Opaque,
```
```suggestion
            .alphaMode = wgpu::CompositeAlphaMode::Auto,
```
(Or you could remove this and presentMode like you did in the sample.)

kainino0x commented on src/library_webgpu.js:
```diff
@@ -2686,6 +2689,11 @@ var LibraryWebGPU = {
+    var viewFormats = {{{ makeGetValue('config', C_STRUCTS.WGPUSurfaceConfiguration.viewFormats, '*') }}};
+    assert(viewFormatCount === 0 && viewFormats === 0, "TODO: Support viewFormats.");
+    var alphaMode = {{{ gpu.makeGetU32('config', C_STRUCTS.WGPUSurfaceConfiguration.alphaMode) }}};
+    assert({{{ gpu.CompositeAlphaMode.Opaque }}} === alphaMode, "TODO: Support alphaMode.");
```
Ah... we should allow `Auto` as well, since it is the default in the C++ header, and it always does the same thing as `Opaque` in Emscripten anyway. [1]

```suggestion
    assert(alphaMode === {{{ gpu.CompositeAlphaMode.Auto }}} ||
      alphaMode === {{{ gpu.CompositeAlphaMode.Opaque }}},
      "TODO: Support WGPUCompositeAlphaMode_Premultiplied.");
```

beaufortfrancois replied:
FYI The default is "Opaque" not "Auto" in https://source.chromium.org/chromium/chromium/src/+/main:out/linux-Debug/gen/third_party/dawn/include/dawn/webgpu_cpp.h;l=2965;drc=6443d1ded6b6a1b62c65099d5014ab240b68946f

beaufortfrancois replied:
I've started https://dawn-review.googlesource.com/c/dawn/+/190080 to address this issue and will update Emscripten.

kainino0x commented on src/library_webgpu.js:
```diff
@@ -99,6 +99,9 @@ wgpu${type}Release: (id) => WebGPU.mgr${type}.release(id),`;
       DeviceLost: 2,
       Unknown: 3,
     },
+    CompositeAlphaMode: {
```
```suggestion
    CompositeAlphaMode: {
      Auto: 0,
```

kainino0x reviewed (approved):
LGTM with one last thing

beaufortfrancois commented:
@sbc100 Is there any chance we can publish next Emscripten release (3.1.61 I believe) when Chrome 126 hits stable channel (June 10th) so that developers don't have to use the ToT version of Emscripten to use the Surface API in WebGPU?

sbc100 commented:
Sure, we can basically cut releases as we please.

sbc100 commented:
(assuming the tree is green and there are no known issues)

beaufortfrancois commented:
> Sure, we can basically cut releases as we please.

That's great to hear!

beaufortfrancois commented:
@sbc100 Can we schedule a new release next week?

sbc100 commented:
We are working on right now.  But we have hit a few snags.

beaufortfrancois commented:
Thanks!

beaufortfrancois commented:
Thanks for releasing https://github.com/emscripten-core/emscripten/releases/tag/3.1.61 @sbc100!
